crates/bevy_pbr/src/meshlet/asset.rs

```diff
@@ -68,84 +70,141 @@ pub struct MeshletBoundingSpheres {
-            .await?;
+        reader: &'a mut dyn Reader,
+        _settings: &'a (),
+        _load_context: &'a mut LoadContext<'_>,
+    ) -> Result<MeshletMesh, MeshletMeshSaveOrLoadError> {
+        // Load and check asset version
+        let version = async_read_u64(reader).await?;
+        if version != MESHLET_MESH_ASSET_VERSION {
+            return Err(MeshletMeshSaveOrLoadError::WrongVersion { found: version });
+        }
 
-        let mut bytes = Vec::new();
-        let mut sync_writer = FrameEncoder::new(&mut bytes);
-        bincode::serialize_into(&mut sync_writer, asset.get())?;
-        sync_writer.finish()?;
-        writer.write_all(&bytes).await?;
+        // Load asset metadata
+        let worst_case_meshlet_triangles = async_read_u64(reader).await?;
+        let compressed_asset_data_len = async_read_u64(reader).await? as usize;
+
+        // Load compressed asset data
+        let mut compressed_asset_data = Vec::with_capacity(compressed_asset_data_len);
+        reader.read_to_end(&mut compressed_asset_data).await?;
+
+        // Convert compressed data back to an asset
+        let reader = &mut FrameDecoder::new(Cursor::new(compressed_asset_data));
+        let vertex_data = read_slice(reader)?;
+        let vertex_ids = read_slice(reader)?;
+        let indices = read_slice(reader)?;
+        let meshlets = read_slice(reader)?;
+        let bounding_spheres = read_slice(reader)?;
+
+        Ok(MeshletMesh {
+            worst_case_meshlet_triangles,
+            vertex_data,
+            vertex_ids,
+            indices,
+            meshlets,
+            bounding_spheres,
+        })
+    }
 
-        Ok(())
+    fn extensions(&self) -> &[&str] {
+        &["meshlet_mesh"]
     }
 }
 
 #[derive(thiserror::Error, Debug)]
 pub enum MeshletMeshSaveOrLoadError {
     #[error("expected asset version {MESHLET_MESH_ASSET_VERSION} but found version {found}")]
     WrongVersion { found: u64 },
-    #[error("failed to serialize or deserialize asset data")]
-    SerializationOrDeserialization(#[from] bincode::Error),
     #[error("failed to compress or decompress asset data")]
     CompressionOrDecompression(#[from] lz4_flex::frame::Error),
     #[error("failed to read or write asset data")]
     Io(#[from] std::io::Error),
 }
 
-async fn read_u64(reader: &mut dyn Reader) -> Result<u64, bincode::Error> {
+async fn async_read_u64(reader: &mut dyn Reader) -> Result<u64, std::io::Error> {
     let mut bytes = [0u8; 8];
     reader.read_exact(&mut bytes).await?;
     Ok(u64::from_le_bytes(bytes))
 }
+
+fn read_u64(reader: &mut dyn Read) -> Result<u64, std::io::Error> {
+    let mut bytes = [0u8; 8];
+    reader.read_exact(&mut bytes)?;
+    Ok(u64::from_le_bytes(bytes))
+}
+
+fn write_slice<T: Pod>(
+    field: &[T],
+    writer: &mut dyn Write,
+) -> Result<(), MeshletMeshSaveOrLoadError> {
+    writer.write_all(&(field.len() as u64).to_le_bytes())?;
+    writer.write_all(bytemuck::cast_slice(field))?;
```

Comment:
Might be worth asserting the that the [target is little endian](https://doc.rust-lang.org/reference/conditional-compilation.html#target_endian) rather than silently storing incompatible data. Another option is using the byteorder crate to handle big endian CPUs, but I'm a bit skeptical anyone would try running Bevy on one

Reply:
Yeah I was thinking that, I'll add a little-endian assert when writing the file, and when loading the plugin.